### PR TITLE
Output sqllogictests with arrow display rather than CSV writer

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?
re https://github.com/apache/arrow-datafusion/issues/4460

# Rationale for this change

A follow up from https://github.com/apache/arrow-datafusion/pull/4547 which parsed CSV writer output

After https://github.com/apache/arrow-datafusion/pull/4547, the sqllogictest  code writes RecordBatches to csv and then reparses it. This can lead to misleading results if the data had ' ' in it, for example.

# What changes are included in this PR?
1. Generate `Vec<Vec<String>>` directly
2. Use `arrow::util::display` for string display rather than the writer. This is what `pretty_format_batches` uses underneath so should be more consistent with existing tests
3. Check that all output RecordBatches have the same schema

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/tests/sqllogictests/test_files/aggregate.slt [hunk 1]
```diff
@@ -52,7 +52,7 @@ SELECT var_pop(c2) FROM aggregate_test_100
 query R
 SELECT var_pop(c6) FROM aggregate_test_100
 ----
-2.615633434202189e37
+26156334342021890000000000000000000000
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/aggregate.slt:55]: these differences are due the difference in `arrow::util::display::array_value_to_string` and the CSV writer. I think the changes are an improvement

File: datafusion/core/tests/sqllogictests/src/normalize.rs [hunk 2]
```diff
@@ -15,61 +15,86 @@
...
-            v
+    let s = match col.data_type() {
+        DataType::Utf8 | DataType::LargeUtf8 if s.is_empty() => {
+            // All empty strings are replaced with this value
+            "(empty)".to_string()
         }
-    })
+        // no other normalization
+        _ => s,
+    };
```
Comment (xudong963) [inline datafusion/core/tests/sqllogictests/src/normalize.rs:97]: I think the following is more simple and reduces assignment.
```suggestion
    let mut s = arrow::util::display::array_value_to_string(col, row)
        .map_err(DFSqlLogicTestError::Arrow)?;

    // apply subsequent normalization depending on type

    if matches!(col.data_type(), DataType::Utf8 | DataType::LargeUtf8) && s.is_empty() {
        // All empty strings are replaced with this value
        s = "(empty)".to_string()
    }
```

**Review by xudong963 — APPROVED** [review]
Very clear, thanks @alamb

File: datafusion/core/tests/sqllogictests/src/normalize.rs [hunk 3]
```diff
@@ -15,61 +15,86 @@
...
+        return Ok("NULL".into());
+    }
 
-                // no normalization on this type
-                _ => array.clone(),
-            }
-        })
-        .collect();
+    // Convert to normal string representation
+    let s = arrow::util::display::array_value_to_string(col, row)
```
Comment (xudong963) [inline datafusion/core/tests/sqllogictests/src/normalize.rs:85]: I think the following is more simple and reduces assignment. (85~97)
```suggestion
    let mut s = arrow::util::display::array_value_to_string(col, row)
        .map_err(DFSqlLogicTestError::Arrow)?;

    // apply subsequent normalization depending on type

    if matches!(col.data_type(), DataType::Utf8 | DataType::LargeUtf8) && s.is_empty() {
        // All empty strings are replaced with this value
        s = "(empty)".to_string()
    }
```

Comment (alamb) [reply datafusion/core/tests/sqllogictests/src/normalize.rs]: Good idea -- done in d0e7db9f8

**Pull request merged**